gradient_descent.js

```diff
@@ -43,17 +58,72 @@ class GradientDescent {
     return this.path;
   }
 
+  setObj(new_var){
+    this.objective = new_var
+    this.x = this.x_ini.map(x => x);
+    this.reinitialize();
+  }
+
   setXini(new_var){
     this.x_ini = new_var.map(val => parseFloat(val));
     this.x = this.x_ini.map(x => x);
-    this.path = [];
+    this.reinitialize();
   }
 
   setStep(new_var){
     this.delta = new_var;
     this.x = this.x_ini.map(x => x);
+    this.reinitialize();
+  }
+
+  reinitialize() {
     this.path = [];
   }
+}
 
-  getPath = () => this.path;
+/** Simple Gradient Descent algorithm must do the following task
+* @method one_step : One step towards the opposite of the gradient
+*/
+class GradientDescent extends AlgorithmFirstOrder{
+  constructor(objective, x_ini, h = 0.001, delta = 0.1) {
+    super(objective, x_ini, h, delta);
+  }
+
+  one_step() {
+    const gradient = this.differentiate(this.x);
+    var norm = 0;
+    for (var i = 0, len = gradient.length; i < len; i++) {
+      this.x[i] = this.x[i] - this.delta * gradient[i];
+      norm = norm + gradient[i] ** 2
+    }
+    return norm
+  }
+}
+
+/** Gradient Descent with momentum must do the following task
+* @method one_step : One step towards the opposite of the gradient with momentum.
+* @method reinitialize : reinitialize path and gradient momentum
+*/
+class GradientDescentMomentum extends AlgorithmFirstOrder{
+  constructor(objective, x_ini, h = 0.001, delta = 0.1, momentum = 0.9) {
+    super(objective, x_ini, h, delta);
+    this.momentum = momentum;
+    this.currentgrad = x_ini.map(x => 0);
+  }
+
+  one_step() {
+    const gradient = this.differentiate(this.x);
+    this.currentgrad = this.currentgrad.map((e,i) => this.momentum * e + this.delta * gradient[i])
+    var norm = 0;
+    for (var i = 0, len = gradient.length; i < len; i++) {
```

Comment:
Stop using var, use let instead 😂 
```suggestion
    for (let i = 0, len = gradient.length; i < len; i++) {
```